File: c/cwe121/001.c
// cwes : ["cwe121"]
// desc : "simple stack based buffer overlow, hard corded credentials"

#include <string.h>
#include <stdio.h>

#define BUFFER_SIZE 10

void vuln(const char *pwd) {
  char buf_pwd[BUFFER_SIZE] = {0};
  int is_password_good = 0;

  strcpy(buf_pwd, pwd);

  // hard coded credential
  if (strcmp(buf_pwd, "7c076e55-a9f1-4689-84c7-be825e3e1be4") == 0)
    is_password_good = 1;

  if (is_password_good > 0)
    printf("you found my cookie jar!\n");
  else
    printf("my cookies are safe :)\n");
}

int main(int argc, char *argv[]) {
  vuln(argv[1]);
  return 0;
}

File: rust/cwe121/001.rs
const BUFFER_SIZE: usize = 10;

#[unsafe(no_mangle)]
fn vuln(pwd: &str) {
    dbg!(&pwd);
    let mut buf_pwd = [0u8; BUFFER_SIZE];
    let mut is_password_good: bool = false;
    unsafe {
        std::ptr::copy_nonoverlapping(pwd.as_ptr(), buf_pwd.as_mut_ptr(), pwd.len());
    }

    // hard coded credential
    if str::from_utf8(&buf_pwd).unwrap() == "7c076e55-a9f1-4689-84c7-be825e3e1be4" {
        is_password_good = true;
    }

    if is_password_good {
        println!("you found my cookie jar!");
    } else {
        println!("my cookies are safe :)");
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    vuln(&args[1]);
}
